### PR TITLE
feat: SearchBox

### PR DESCRIPTION
**Pull request by PLNech**
SearchBox components:
- `SearchBoxView`
- `SearchBoxViewModel`
- connectors  

Demos:
- Factorize `DemoActivity`
- `SearchOnSubmitDemo`
- `SearchEditTextDemo`

### CODE REVIEW

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import kotlin.properties.Delegates
+
+/**
+ * A ViewModel for SearchBoxView implementations, holding their [query].
+ *
+ * It lets you define an arbitrary number of listeners,
+ * for both [text change][changeListeners] and [submit][submitListeners] events.
+ */
+class SearchBoxViewModel {
+    var query by Delegates.observable<String?>(null) { _, old, new ->
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt]: Always put a line after the class declaration.

Also Uuse `oldValue` and `newValue` like everywhere else. It is the suggested name given by the IDE autocomplete.

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import kotlin.properties.Delegates
+
+/**
+ * A ViewModel for SearchBoxView implementations, holding their [query].
+ *
+ * It lets you define an arbitrary number of listeners,
+ * for both [text change][changeListeners] and [submit][submitListeners] events.
+ */
+class SearchBoxViewModel {
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt]: Should be public

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        if (old != new) {
+            changeListeners.forEach { it(new) }
+        }
+    }
+
+    fun submit() {
+        submitListeners.forEach { it(query) }
+    }
+
+    var changeListeners: List<(String?) -> Unit> = listOf()
```
Comment (q-litzler) [inline core/src/commonMain/kotlin/com/algolia/instantsearch/core/searchbox/SearchBoxViewModel.kt]: Use the norm we agreed upon:

`onQueryChanged`
`onQuerySubmitted`

File: demo/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -60,7 +60,15 @@
             android:parentActivityName=".home.HomeDemo" />
 
         <activity
-            android:name=".filter.toggle.FilterToggleDefault"
+            android:name=".filter.toggle.FilterToggleDefaultDemo"
+            android:parentActivityName=".home.HomeDemo" />
+
+        <activity
+            android:name="com.algolia.instantsearch.demo.searchbox.SearchOnSubmitDemo"
```
Comment (q-litzler) [inline demo/src/main/AndroidManifest.xml]: Name the activity `SearchBoxOnSubmitDemo`

File: demo/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -60,7 +60,15 @@
...
+            android:parentActivityName=".home.HomeDemo" />
+
+        <activity
+            android:name="com.algolia.instantsearch.demo.searchbox.SearchEditTextDemo"
```
Comment (q-litzler) [inline demo/src/main/AndroidManifest.xml]: Name the activity `SearchBoxOnEditTextDemo`

File: demo/src/main/kotlin/com/algolia/instantsearch/demo/Demo.kt [hunk 6]
```diff
@@ -97,13 +97,19 @@ fun AppCompatActivity.configureRecyclerView(
...
-    searchView: SearchView
+    searchView: SearchView,
+    searchAsYouType: Boolean = true
 ) {
     searchView.also {
         val initialHintText = searchView.queryHint.toString()
         val hintIcon = ContextCompat.getDrawable(this, R.drawable.ic_search_hint)!!
-        it.isSubmitButtonEnabled = false
-        it.isFocusable = true
+        if (searchAsYouType) {
```
Comment (q-litzler) [inline demo/src/main/kotlin/com/algolia/instantsearch/demo/Demo.kt]: If this is part of the way has to configure the searchView for it to work properly, this piece of code should live in the activity where it is demoed. We should not be hiding it.

**Comment by PLNech** [comment]
> no abstract `Activity`  

Trusting you on this and reverted in 0ace696, but I can't find a good resource explaining why it's a bad pattern in general. I find [this](https://blog.edfora.com/commonly-used-anti-patterns-in-android-a0c6fe070678) offering an alternative without explaining why, while [Clean Android Code](https://medium.com/spring-media-techblog/keep-your-droid-clean-activities-and-fragments-da1799fe1e7a) explicitly recommends it. If you have an authoritative reference on the matter, I'd love to read it :)

**Comment by PLNech** [comment]
> Supporting anything other than appcompat

I have the impression that if at least some of our users chose not to use AppCompat, we should still support them.  
However as I can't find anyone online with such a good reason, and as quoting [AppCompatEditText's docs](https://developer.android.com/reference/android/support/v7/widget/AppCompatEditText.html) such an AppCompat widget `will automatically be used when you use EditText in your layouts and the top-level activity / dialog is provided by appcompat`, I guess we can safely ignore non-appcompat views unless users start asking for it!

File: demo/src/main/kotlin/com/algolia/instantsearch/demo/Demo.kt (hunk 6 above)
Comment (PLNech) [reply demo/src/main/kotlin/com/algolia/instantsearch/demo/Demo.kt]: It's not more necessary than other calls to `configureSearchView`: although the UI is better configured as such, it's purely cosmetic and the demo works without displaying a Submit button:
cosmetic

**Comment by PLNech** [comment]
> Let's demo only searchView

I don't see the value in showing a single kind of `View` instead of showcasing the customizability of IS within our demos. If we demonstrate `two different modes (onSubmit vs onQueryChange)`, I'd rather see us take it as an opportunity to show two kinds of UIs :)

This being said I'll rename the second demo `SearchBoxAsYouTypeDemo` to highlight the mode it demonstrates, making the `EditText` UI a minor detail of it!

**Comment by PLNech** [comment]
> Don't forget public visibility modifier

-> Does this apply to `interface`s? I'd say yes, in which case I'll add the missing ones to `FilterListView` and `FilterPresenter`

**Pull request merged**